streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -299,18 +300,22 @@ public void resume() {
     }
 
     /**
-     * An active task is processable if its buffer contains data for all of its input source topic partitions
+     * An active task is processable if its buffer contains data for all of its input
+     * source topic partitions, or if it is enforced to be processable
      */
-    public boolean isProcessable() {
-        if (partitionGroup.allPartitionsBuffered()) {
+    boolean isProcessable(final long now) {
+        if (enforceProcessing) {
+            return true;
+        } else if (partitionGroup.allPartitionsBuffered()) {
             return true;
-        } else if (partitionGroup.numBuffered() > 0 && --waits < 0) {
+        } else if (partitionGroup.numBuffered() > 0 && now - lastEnforcedProcessingTime > maxTaskIdleMs) {
             taskMetrics.taskEnforcedProcessSensor.record();
-            waits = WAIT_ON_PARTIAL_INPUT;
+            lastEnforcedProcessingTime = now;
+            enforceProcessing = true;
```

Comment:
Overall, the enforced-processing algorithm is unclear to me.

* Following on @mjsax 's comment, it seems strange to set this right before we return true anyway. Note that this is currently the only place we set `enforceProcessing` to true.

* Also, it still seems to me that `maxTaskIdleMs` should count from the last time we process at all, not the last time we forced processing (same basic scenario I pointed out last time). Is this right?

Reply:
Here is the rationale of this logic:

1. Once we decide to `enforceProcessing`, we will continue enforcing until we got new data enqueued and all the buffer become full, in this case we will in `normalProcessing` state.
2. But we will update the `lastEnforcedProcessingTime` the last time we decide to start enforce processing.
3. We know that once we decide to enforce processing, we will always process immediately as there are indeed some data buffered already.

So the logic above sets `lastEnforcedProcessingTime` at the time we decide to "turn it on", and only "turn it off" during records enqueuing and all buffers contain some data. And hence we will first check `enforceProcessing`: if it is true we just continue enforce processing.

LGTY?

Reply:
I see. Thanks for explaining.

It still seems like the `enforceProcessing` variable isn't strictly necessary, it just saves calls to `partitionGroup.allPartitionsBuffered()`, `partitionGroup.numBuffered()`, and the comparison `now - lastEnforcedProcessingTime > maxTaskIdleMs`. These are all just cached field lookups, though, so I don't know if the performance boost is worth the algorithmic complexity.

Regarding `lastEnforcedProcessingTime`, consider this scenario.
```
maxTaskIdleMs := 2
lastEnforcedProcessingTime := 0 (init)
now=0; allPartitionsBuffered=true => isProcessable:=true
now=1; allPartitionsBuffered=true => isProcessable:=true
now=2; allPartitionsBuffered=true => isProcessable:=true
now=3; allPartitionsBuffered=false => 
  // 3 - 0 > 2, so we force processing
  lastEnforcedProcessingTime:=3
  isProcessable:=true
```

Two things to note here:
1. the expression should probably be `now - lastEnforcedProcessingTime >= maxTaskIdleMs` (with `>=` instead of `>`), otherwise you'll wait at least one extra ms _beyond_ the purported "max task idle time".
2. In the scenario above, we said we want to wait *2 ms* before forcing processing, but we actually force processing *immediately*. To fix this, we should be comparing against `lastProcessingTime`, which we should set every time we process.

Reply:
Good point, I will try to address this along with @mjsax 's other comment:

> Good point. Now I am wondering, if we should set lastEnforcedProcessingTime = Long.MAX_VALUE, too, when we set enforceProcessing == false when adding records to the buffers?

Reply:
Thanks. One final thought about whether the `enforceProcessing` optimization is worth it. It might be a good idea to benchmark it without the optimization, since branch prediction *should* eliminate any overhead from the checks on rarely used branches.

Reply:
@vvcephei @mjsax this turns out to be harder than I thought. The tricky thing is that we originally want to 1) record the sensor only for the first time when we transit to "enforced processing" state, and 2) start the idleness timer only for the first time when we do not have all buffered but some buffered.

I tried to even implement a StreamTask State just like KafkaStreams and StreamThread, but that turns out to not be so elegant as well. So what I ended up now is this: we will record the sensor whenever we enforce processing, either for the first time or not, and hence we will only update idleStartTime once, and reset it whenever we have all buffered.

LMK WDYT.

Reply:
> So what I ended up now is this: we will record the sensor whenever we enforce processing, either for the first time or not, and hence we will only update idleStartTime once, and reset it whenever we have all buffered.

Sounds good to me.